experimental/lib/Transforms/Speculation/HandshakeSpeculation.cpp

```diff
@@ -127,63 +128,103 @@ static void markPathToCommits(llvm::DenseSet<Operation *> &markedPath,
+  for (OpOperand &succOpOperand : specOp.getDataOut().getUses()) {
+    queue.push(
+        std::tuple<const OpOperand &, CommitBranchList>(succOpOperand, {}));
+  }
+
+  MLIRContext *ctx = &getContext();
+  OpBuilder builder(ctx);
+
+  // Perform BFS
+  while (!queue.empty()) {
+    auto [currOpOperand, commitBranchList] = queue.front();
+    queue.pop();
+    Operation *currOp = currOpOperand.getOwner();
+
+    if (auto commitOp = dyn_cast<handshake::SpecCommitOp>(currOp)) {
+      // We replicate branches only if the traversal reaches a commit.
+      // Because sometimes a path of branches does not reach a commit unit.
+      // (appropriate) path when we traverse other branches.
+      Value ctrlSignal = specOp.getCommitCtrl();
+      for (auto [valueForSpecTag, branchOp, branchDir] : commitBranchList) {
+        // Replicate a branch in the control path and use new control signal.
+        // To do so, a structure of two connected branches is created.
+        // A speculating branch first discards the condition in case that
+        // the data is not speculative. In case it is speculative, a new branch
+        // is created that replicates the current branch.
+
+        builder.setInsertionPointAfterValue(ctrlSignal);
+
+        // The speculating branch will discard the branch's condition token if
+        // the branch output is non-speculative. Speculative tag of the token is
+        // currently implicit, so the branch input itself is used at the IR
+        // level.
+        auto branchDiscardNonSpec =
+            builder.create<handshake::SpeculatingBranchOp>(
+                branchOp.getLoc(), valueForSpecTag /* specTag */,
+                branchOp.getConditionOperand());
+        inheritBB(specOp, branchDiscardNonSpec);
+
+        // The replicated branch directs the control token based on the path the
+        // speculative token took
+        auto branchReplicated = builder.create<handshake::ConditionalBranchOp>(
+            branchDiscardNonSpec->getLoc(),
+            branchDiscardNonSpec.getTrueResult() /* condition */,
+            ctrlSignal /* data */);
+        inheritBB(specOp, branchReplicated);
+
+        ctrlSignal = branchReplicated->getResult(branchDir);
       }
-    }
-  } else {
-    // Continue Traversal
-    for (OpResult res : currOp->getResults()) {
-      for (OpOperand &dstOpOperand : res.getUses()) {
-        routeCommitControl(markedPath, ctrlSignal, dstOpOperand);
+      // Connect commit to the correct control signal and end traversal
+      commitOp.setOperand(1, ctrlSignal);
+    } else if (auto branchOp =
+                   dyn_cast<handshake::ConditionalBranchOp>(currOp)) {
+      // Follow the two branch results with a different control signal
+      for (unsigned i = 0; i <= 1; ++i) {
+        // Copy the current list. Can be optimized by using a data structure
+        // with reference
+        auto newList(commitBranchList);
```

Comment:
add explicit comment about storing which branches to replicate ( I didn't know what the list was for)